### ISSUE TITLE
How to Relinearize. SystemError: <method 'relinearize' of 'Pyfhel.Pyfhel.Pyfhel' objects> returned a result with an error set

### ISSUE THREAD

**Issue by vinegarg** [ibarrond/Pyfhel#58]
Hi,I recently got my hands-on with this amazing library.But i got some problems when i applied logistic regression model with Stochastic gradient descent on encrypted data(Dataset:141rows*3cols). when updating the weights until the fifth time,the size of W*x getting very large.Maybe I should use relinKeyGen(int bitCount, int size) and relinearize(PyCtxt ctxt) to reduce its size down to 2.

1.Before i didn't use relinearize(), the sizes showed as followings.
  ```python
       <Pyfhel Ciphertext,encoding=FRACTIONAL,size=3>
       <Pyfhel Ciphertext,encoding=FRACTIONAL,size=12>
       <Pyfhel Ciphertext,encoding=FRACTIONAL,size=48>
       <Pyfhel Ciphertext,encoding=FRACTIONAL,size=192>
       <Pyfhel Ciphertext,encoding=FRACTIONAL,size=768>
```
And then the Server went wrong.

2.when initializing relinKeyGen(20,6) and using relinearize(),another problem occured.
```python
SystemError: <method 'relinearize' of 'Pyfhel.Pyfhel.Pyfhel' objects> returned a result with an error set
```

The key codes are as follow.The only difference about situation 1&2 is " self.pyfhel.relinearize(hy) ".   
```python
def dot_product(self,a,theta):
       sum=self.pyfhel.encryptFrac(0)
       for i in range(len(a)):
            sum=self.pyfhel.add(sum,self.pyfhel.multiply(a[i],theta[i],in_new_ctxt=True),in_new_ctxt=True)
       print("dot:",sum)
       return sum

def sigmoid():
       pass
            
def BGD(self,X_enc,y_enc):
        m,n=X_enc.shape
        theta=np.ones((n,1))
        theta=self.encrypt_theta(theta)
        iters=100
        for i in range(iters):
            dataIndex=list(range(m))
            for j in range(m):
                alpha = 4.0/(1.0+i+j)+0.01
                alpha_enc=self.pyfhel.encryptFrac(alpha)
                randIndex=int(random.uniform(0,m))
                hy=self.dot_product(X_enc[randIndex],theta)
                **self.pyfhel.relinearize(hy) //SystemError:：<method 
                'relinearize' of 'Pyfhel.Pyfhel.Pyfhel' objects> returned a 
                 result with an error set**
                X_sigmoid=self.sigmoid(hy)
                error=self.pyfhel.sub(X_sigmoid,y_enc[randIndex])
                error_alpha=self.pyfhel.multiply(error,alpha_enc)
                x_error_alpha=self.mul_fra(X_enc[randIndex],error_alpha)
                theta=self.sub_array(theta,x_error_alpha)
                del(dataIndex[randIndex])
        return theta
```
My question:
1.problem ——the size of dot.product getting large,can we use the function relinKeyGen() and using relinearize() to solve it?
2.If it's,how to use the function relinKeyGen() and relinearize()  ,even the Args relinKeyGen().I couldn't find any example in github  -_- so sad.
    If it's not,what should I do next?
    Looking forward to your reply soon.Thankkkkkkk so much!

**Comment by ibarrond** [comment]
Hi! thanks for the positive feedback. You're right! Pyfhel doesn't have an example using the relinearization keys. We should amend that mistake!

I'm having some issues reading your code. could you please format it properly so that for loops and indentation is correct?

Regarding Relinearization, I recommend you to try applying it right after each multiplication, to keep things as small as possible. The two parameters for relinKeys are a bit obscure, and sadly I don't remember much about them. Please share any experiments you do with them! Newer versions of the backend library SEAL have removed them; It'll be simpler after I finish with #25 .

In short, try relinearization after each multiplication and play a bit with the parameters. Share you results here!

**Comment by ibarrond** [comment]
Hi, can you share the full code? I'd like to reproduce your exact error, and I nee the `contextGen` parameter for this.

**Comment by ibarrond** [comment]
I'd say the real error you're getting is (which you should find on top of your error message):

```bash
ValueError: not enough evaluation keys
[...]
```

**Comment by ibarrond** [comment]
Solved it! When generating relinearization Keys using `relinKeyGen(a, b)`, the bitcount (parameter `b`) must be bigger than the `size` of the PyCtxt you're gonna relinearize. Otherwise a `ValueError: not enough evaluation keys` will be thrown.

**Comment by ibarrond** [comment]
The commit bd16db37644fe4b3eb01f060211397bbb86bc4cf adds a nice demo to show relinearization, as well as some nice printing of Ciphertexts and Pyfhel objects. Hopefully it is clear enough now!

**Comment by vinegarg** [comment]
oh,I'm so sorry.I was too busy to forget this message.Thanks for your answer.Maybe I'll try again.I'll show you the codes.But I'm a code junior.Please please give me some advices.Thank you!!!
import numpy as np
from Pyfhel.util import ENCODING_t
import pandas as pd
import random
from sklearn import preprocessing
# Local module
from Pyfhel import PyCtxt, Pyfhel, PyPtxt
import gc
import random

class HELR(object):
    
    def __init__(self):
        self.pyfhel = Pyfhel()
        self.pyfhel.contextGen(p=65537, m=8192, base=2, sec=192, flagBatching=True)
        self.pyfhel.keyGen()
        self.pyfhel.relinKeyGen(10, 6)
    
    def encrypt_a(self,a):
        a_enc=[]
        for i in range(a.shape[0]):
            row_val=[]
            for j in range(a.shape[1]):
                q_enc=self.pyfhel.encryptFrac(a[i][j])
                row_val.append(q_enc)
                del q_enc
                gc.collect()
            a_enc.append(row_val)
        a_enc = np.asarray(a_enc)
        return a_enc
    
    def encrypt_theta(self,theta):
        theta_enc=[]
        for j in range(theta.shape[0]):
            q_enc=self.pyfhel.encryptFrac(theta[j][0])
            theta_enc.append(q_enc)
            del q_enc
            gc.collect()
        theta_enc=np.asarray(theta_enc)
        theta_enc = theta_enc.reshape(theta.shape[0])
        return theta_enc
    
    def encrypt_y(self,y):
        y_enc=[]
        for i in range(y.shape[0]):
            q_enc=self.pyfhel.encryptFrac(y[i][0])
            y_enc.append(q_enc)
            del q_enc
            gc.collect()                             
        y_enc=np.asarray(y_enc)
        y_enc = y_enc.reshape(141)
        return y_enc
            
    def dot_product(self,a,theta):
        sum=self.pyfhel.encryptFrac(0)
        for i in range(len(a)):
            sum=self.pyfhel.add(sum,self.pyfhel.multiply(a[i],theta[i],in_new_ctxt=True),in_new_ctxt=True)
        print("dot:",sum)
        return sum
    
    def sigmoid(self,sum):
        ar=self.pyfhel.square(sum)
        arr=self.pyfhel.multiply_plain(ar,self.pyfhel.encodeFrac(-0.004),in_new_ctxt=True)
        del ar
        gc.collect()
        arrr=self.pyfhel.add_plain(arr,self.pyfhel.encodeFrac(0.197),in_new_ctxt=True)
        del arr
        gc.collect()
        arrrr=self.pyfhel.multiply(arrr,sum)
        del arrr
        gc.collect()
        arrrrr=self.pyfhel.add_plain(arrrr,self.pyfhel.encodeFrac(0.5),in_new_ctxt=True)
        return arrrrr
    
    def dot_product_predict(self,a,theta):
        res=[]
        for i in range(a.shape[0]):
                sum=self.pyfhel.encryptFrac(0)
                for k in range(a.shape[1]):
                    sum=self.pyfhel.add(sum,self.pyfhel.multiply(a[i][k],theta[k],in_new_ctxt=True),in_new_ctxt=True)
                res.append(sum)
                del sum
                gc.collect()                           
        res=np.asarray(res)
        del a,theta
        gc.collect()
        return res
    
    def sqr1(self,res):
        a_sqr=[]
        for i in range(res.shape[0]):
            q_enc=self.pyfhel.square(res[i])
            a_sqr.append(q_enc)
            del q_enc
            gc.collect()
        a_sqr=np.asarray(a_sqr)
        del res
        gc.collect()
        return a_sqr
    
    def sqr2(self,a_sqr):
        ar_sqr=[]
        for i in range(a_sqr.shape[0]):
            q_enc=self.pyfhel.multiply_plain(a_sqr[i],self.pyfhel.encodeFrac(-0.004),in_new_ctxt=True)
            ar_sqr.append(q_enc)
            del q_enc
            gc.collect()                                
        ar_sqr=np.asarray(ar_sqr)
        del a_sqr
        gc.collect()
        return ar_sqr
    
    def sqr3(self,ar_sqr):
        arr_sqr=[]
        for i in range(ar_sqr.shape[0]):
            q_enc=self.pyfhel.add_plain(ar_sqr[i],self.pyfhel.encodeFrac(0.197),in_new_ctxt=True)
            arr_sqr.append(q_enc)
            del q_enc
            gc.collect()                             
        arr_sqr=np.asarray(arr_sqr)
        del ar_sqr
        gc.collect()
        return arr_sqr
    
    def sqr4(self,arr_sqr,res):
        arrr_sqr=[]
        for i in range(arr_sqr.shape[0]):
            q_enc=self.pyfhel.multiply(arr_sqr[i],res[i])
            arrr_sqr.append(q_enc)
            del q_enc
            gc.collect()                                    
        arrr_sqr=np.asarray(arrr_sqr)
        del arr_sqr,res
        gc.collect()
        return arrr_sqr
    
    def sqr5(self,arrr_sqr):
        X_sigmoid=[]
        for i in range(arrr_sqr.shape[0]):
            q_enc=self.pyfhel.add_plain(arrr_sqr[i],self.pyfhel.encodeFrac(0.5),in_new_ctxt=True)
            X_sigmoid.append(q_enc)
            del q_enc
            gc.collect()                                
        X_sigmoid=np.asarray(X_sigmoid)
        del arrr_sqr
        gc.collect()
        return X_sigmoid
    
    def sigmoid_predict(self,X_enc,theta_true):
        res=self.dot_product(X_enc,theta_true)
        X_sqr=self.sqr1(res)
        Xr_sqr=self.sqr2(X_sqr)
        del X_sqr
        gc.collect()
        Xrr_sqr=self.sqr3(Xr_sqr)
        del Xr_sqr
        gc.collect()
        Xrrr_sqr=self.sqr4(Xrr_sqr,res)
        del res,Xrr_sqr
        gc.collect()
        X_sigmoid_predict=self.sqr5(Xrrr_sqr)
        del Xrrr_sqr
        gc.collect()
        return X_sigmoid_predict
    
    def decrypt_a(self,a):
        dr=[]
        for i in range(a.shape[0]):
            dr.append(self.pyfhel.decryptFrac(a[i]))
        dr=np.asarray(dr)
        return dr
    
    def BGD(self,X_enc,y_enc):
        m,n=X_enc.shape
        theta=np.ones((n,1))
        theta=self.encrypt_theta(theta)
        iters=3
        for i in range(iters):
            dataIndex=list(range(m))
            for j in range(m):
                alpha = 4.0/(1.0+i+j)+0.01
                alpha_enc=self.pyfhel.encryptFrac(alpha)
                del alpha
                gc.collect()
                randIndex=int(random.uniform(0,m))
                hy=self.dot_product(X_enc[randIndex],theta)
                X_sigmoid=self.sigmoid(hy)
                del hy
                gc.collect()
                error=self.pyfhel.sub(X_sigmoid,y_enc[randIndex])
                del X_sigmoid
                gc.collect()
                error_alpha=self.pyfhel.multiply(error,alpha_enc)
                del error,alpha_enc
                gc.collect()
                x_error_alpha=self.mul_fra(X_enc[randIndex],error_alpha)
                del error_alpha
                gc.collect()
                theta=self.sub_array(theta,x_error_alpha)
                del x_error_alpha
                gc.collect()
                del(dataIndex[randIndex])
        return theta
    
    def sub_array(self,a,b):
        c=[]
        for i in range(3):
            q_enc=self.pyfhel.sub(a[i],b[i])
            c.append(q_enc)
            del q_enc
            gc.collect()
        c=np.asarray(c)
        return c
    
    def mul_fra(self,a,b):
        c=[]
        for i in range(3):
            q_enc=self.pyfhel.multiply(a[i],b)
            c.append(q_enc)
            del q_enc
            gc.collect()
        c=np.asarray(c)
        return c
    
    def predict(self,X_enc,theta_true):
        probability=self.sigmoid_(X_enc,theta_true)
        probability_dec=self.decrypt_a(probability)
        del probability
        gc.collect()
        return [1 if x >=0.5 else 0 for x in probability_dec]
    
    def prohibit(self,predictions,y):
        cnt=0
        for i in range(len(X)):
            if predictions[i] == y[i]:
                cnt+=1
        return cnt
if __name__ == "__main__":
    he=HELR()
    url='/Users/chenxiaolu/Desktop/CA19_125_Data.txt'
    df = pd.read_csv(url, header=None, names=['Exam1', 'Exam2', 'Admitted'])
    print(df.head(8))
    print("Data shape:",df.shape)
    try:
        pdData.insert(0, 'Ones', 1) # 写到try...except结构中以防第二次执行时报错
    except:
        pass
    df.insert(0,1,1,1)
    orig_data = df.values
    cols = orig_data.shape[1]
    X = orig_data[:,0:cols-1]
    y = orig_data[:,cols-1:cols]
    y
    X_enc=he.encrypt_a(X)
    print(X_enc)
    y_enc=he.encrypt_y(y)
    print(y_enc)
    theta_true=he.BGD(X_enc,y_enc)
    print(theta_true)
    predictions=he.predict(X_enc,theta_true,y)
    print(predictions)
    cnt=he.prohibit(predictions,y)
    print("正确率：",cnt/len(X)*100,'%')


dot: <Pyfhel Ciphertext, encoding=FRACTIONAL, size=3>
dot: <Pyfhel Ciphertext, encoding=FRACTIONAL, size=12>
dot: <Pyfhel Ciphertext, encoding=FRACTIONAL, size=48>
dot: <Pyfhel Ciphertext, encoding=FRACTIONAL, size=192>
dot: <Pyfhel Ciphertext, encoding=FRACTIONAL, size=768>

and then the service hanged
Thank you!!!

**Comment by ibarrond** [comment]
Maybe the entire code wasn't such a big idea. It is very hard to understand. Please remove it from your previous comment and prepare a small, easy to understand example (~20 lines) that generates the same error you're having. I should be able to execute the example and get the same error you're getting.

Rephrasing what I said above so that it becomes clearer, you should **apply relinearization after every multiplication**. There is a nice demo showing relinearization for multiple multiplications in `examples/Demo_MultDepth_n_relin.py`. Please take a deep look at it. When generating relinearization Keys using `relinKeyGen(bitCount=None, size=None)`, the size parameter must be equal or bigger than the size of the PyCtxt you're gonna relinearize (you can get the size of a ciphertext by printing it). Otherwise a ValueError: not enough evaluation keys will be thrown.

**Comment by vinegarg** [comment]
Sorry to disturb you. I still have a hard time understanding how to design the parameter in the initialed relinKeyGen(, ).Especialy,what's the mean of "Make sure the bitcount is higher than the size of your ciphertexts"?😫

print("4. Relinearization reduces the size of a ciphertext back to 2, thus it is very convenient for deep multiplications")
print("    First we generate relinearization keys.")
print("    Make sure the bitcount is higher than the size of your ciphertexts")
HE.relinKeyGen(1, 1)

HE.relinearize(ctxtMul1) # Equivalent to HE.relinearize(ctxtMul)
                      # and equivalent to  ~ctxtMul1 (without assignment)
print(ctxtMul1)
print("   Good! size increase is solved with relinearization.")


Thanks very much!!!

**Comment by ibarrond** [comment]
You're right! The demo and the documentation of `relinKeyGen` wasn't clear. Hopefully it is clearer now. The `size` argument in `relinKeyGen(birCount, size)` function should get an integer value bigger than the size of the ciphertext you want to relinearize (you can get it by running `ctxt.size()`). The code in the demo now looks like this:

```python
print("4. Relinearization reduces the size of a ciphertext back to 2, thus it is very convenient for deep multiplications")
print("    First we generate relinearization keys.")
print("    * Make sure the relinKey size is higher than the size of the ciphertext you want to relinearize.")
print("    * bigger bitCount means faster relinarization, but also bigger decrease in noiseBudget. Needs to be within [1, 60]")
relinKeySize=5
HE.relinKeyGen(bitCount=1, size=relinKeySize)
print("   We can now relinearize!")
assert relinKeySize >= ctxtMul1.size()

ctxtMul1 = ~ ctxtMul1 # Equivalent to HE.relinearize(ctxtMul)
                      # and equivalent to  ~ctxtMul1 (without assignment)
print(ctxtMul1)
print("   Good! size increase is solved with relinearization.")
```

**Comment by vinegarg** [comment]
Thank you very Much!!!I will try again.Hope I'll get a satisfied output,haha.Thanks again!

**Comment by vinegarg** [comment]
Hello,sorry to disturb u again.if I want to use 'relinearize' to reduce the size of ciphertext in different functions but in the same *.py,what should I do is to initialize the relinkey such as situation 1 at beginning with other encryption parameters and calling self. relinearize() in the functions to reduce size？Or innitialing relinkey(10,3) at different function and then calling relinearize() such as situation 2.
Situation 1：

def __init__(self):
    self.pyfhel = Pyfhel()
    self.pyfhel.contextGen(p=65537, m=8192, base=2, sec=192, intDigits=64, fracDigits = 32)
    self.pyfhel.keyGen()
    **self.pyfhel.relinKeyGen(10,3)**

def dot_product(self,a,theta):
    sum=self.pyfhel.encryptFrac(0)
    for i in range(len(a)):
                     sum=self.pyfhel.add(sum,self.pyfhel.multiply(a[i],theta[i],in_new_ctxt=True),in_new_ctxt=True)
    **self.pyfhel.relinearize(sum)**
    return sum

def sqr1(self,res):
    a_sqr=[]
    for i in range(res.shape[0]):
        q_enc=self.pyfhel.square(res[i])
    **self.pyfhel.relinearize(q_enc)**
    a_sqr.append(q_enc)
    a_sqr=np.asarray(a_sqr)
    return a_sqr

Situation 2：

def __init__(self):
    self.pyfhel = Pyfhel()
    self.pyfhel.contextGen(p=65537, m=8192, base=2, sec=192, intDigits=64, fracDigits = 32)
    self.pyfhel.keyGen()

def dot_product(self,a,theta):
    sum=self.pyfhel.encryptFrac(0)
    for i in range(len(a)):
         sum=self.pyfhel.add(sum,self.pyfhel.multiply(a[i],theta[i],in_new_ctxt=True),in_new_ctxt=True)
    **self.pyfhel.relinKeyGen(10,3)
    self.pyfhel.relinearize(sum)**
    return sum

def sqr1(self,res):
    a_sqr=[]
    for i in range(res.shape[0]):
                     q_enc=self.pyfhel.square(res[i])
                                                                                                **self.pyfhel.relinKeyGen(10,3)
     self.pyfhel.relinearize(q_enc)**
     a_sqr.append(q_enc)
     a_sqr=np.asarray(a_sqr)
     return a_sqr 

Looking forward to your reply.Thank you very much!

**Comment by ibarrond** [comment]
Very good question! I see you already got a nice example working. 

You should initialize the keys **only at the very beginning of your computation**, and use the same keys every time you `relinearize()` (your option 1)

**Comment by vinegarg** [comment]
Oh yeah!I think I've got it!My project run well with relinKeyGen() now.But it didn't decrypt as my expected.😂Maybe it's just as you said，I need more noise budget.![image](https://user-images.githubusercontent.com/48371760/96394686-36dc5080-11f5-11eb-8710-ff9d7af3bf99.png)
    As I reduce p,and increase m，such as contextGen(p=63,m=8192),it works well.But the decrypted θ(weights of logistic regression) doesn't as my expected either.
![image](https://user-images.githubusercontent.com/48371760/96395207-a3a41a80-11f6-11eb-9d76-0e191cf220dc.jpeg)
    Is there any other ideas to get more noise budget？
    You are so patient to answer every small question I put.It gives me more confidence in writing codes.Thank you very much again.

**Comment by ibarrond** [comment]
As convenient as it is for you to copy-paste an image of the code, it is very inconvenient for anyone to reuse or reproduce. Please remove the images from the previous comment and add some properly formatted code

**Comment by ibarrond** [comment]
Anyhow, your question about the noise budget seems to be different from the one about relinearization. Please create a new issue and paste the code there. We can leave this issue as a reminder of **how to relinearize**.

**Comment by vinegarg** [comment]
ok!I'm sorry for That!
